File: packages/scheduling-engine/src/constraints/tool-timeline.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — ToolTimeline Constraint
//  No two machines can use the same physical tool at the
//  same time (§6 bdmestre). Tools are physical objects that
//  must be moved between machines.
//
//  Mode: HARD — physically impossible to violate.
//  The engine MUST respect this constraint and delay or
//  block operations when a tool is in use elsewhere.
// ═══════════════════════════════════════════════════════════

import type { ConstraintName } from '../types/constraints.js';

// ── Internal types ──

/** A booked tool usage time slot on a specific machine */
interface ToolSlot {
  start: number;
  end: number;
  machineId: string;
}

/** Result of checking tool availability */
export interface ToolCheckResult {
  /** Whether the tool is available at the requested time/duration */
  isAvailable: boolean;
  /** Earliest minute the tool can be used (-1 if no room in shift) */
  availableAt: number;
  /** Number of conflicting machines (for multi-instance tools) */
  conflictCount: number;
  /** IDs of machines currently using this tool in the overlap window */
  conflictingMachines: string[];
}

// ── Constraint name constant ──

const CONSTRAINT_NAME: ConstraintName = 'TOOL_TIMELINE';
// Used internally for identification; suppress unused warning
void CONSTRAINT_NAME;

// ── Factory ──

/**
 * Creates a ToolTimeline constraint instance.
 *
 * ToolTimeline prevents two machines from using the same physical tool
 * simultaneously. Each tool defaults to 1 instance (one physical copy).
 * Some tools may have multiple instances (e.g., 2 copies of the same mold),
 * controlled by the `instances` parameter.
 *
 * This is a HARD constraint: tools are physical objects, and using
 * a tool that is physically on another machine is impossible.
 *
 * @example
 * ```ts
 * const tl = createToolTimeline()
 * const check = tl.check('BWI003', 500, 600, 'PRM019')
 * if (check.isAvailable) {
 *   tl.book('BWI003', 500, 600, 'PRM019')
 * }
 * ```
 */
export function createToolTimeline() {
  const timelines: Record<string, ToolSlot[]> = {};

  return {
    /**
     * Check whether a tool is available on a given machine during [start, end).
     * A tool on the SAME machine is always compatible (no conflict with self).
     * For tools with multiple instances, up to `instances` machines can use
     * the tool concurrently.
     *
     * @param toolId    - Tool code (e.g., 'BWI003')
     * @param start     - Absolute start minute
     * @param end       - Absolute end minute
     * @param machineId - Machine requesting the tool
     * @param instances - Number of physical tool copies (default 1)
     * @returns true if the tool is available
     */
    isAvailable(
      toolId: string,
      start: number,
      end: number,
      machineId: string,
      instances?: number,
    ): boolean {
      const slots = timelines[toolId];
      if (!slots) return true;
      const maxInst = instances ?? 1;
      const conflicting = new Set<string>();
      for (const s of slots) {
        if (s.machineId === machineId) continue;
        if (start < s.end && end > s.start) conflicting.add(s.machineId);
      }
      return conflicting.size < maxInst;
    },

    /**
     * Find the next available time for a tool on a given machine.
     * Scans existing bookings and finds the earliest gap that fits.
     *
     * @param toolId    - Tool code
     * @param earliest  - Earliest absolute minute
     * @param duration  - Required duration in minutes
     * @param shiftEnd  - Shift end (absolute minutes)
     * @param machineId - Machine requesting the tool
     * @param instances - Number of physical copies (default 1)
     * @returns Absolute minute to start, or -1 if no room in this shift
     */
    findNextAvailable(
      toolId: string,
      earliest: number,
      duration: number,
      shiftEnd: number,
      machineId: string,
      instances?: number,
    ): number {
      const slots = timelines[toolId];
      if (!slots) return earliest;
      const maxInst = instances ?? 1;
      let candidate = earliest;
      let changed = true;
      let iterations = 0;
      while (changed && iterations < 1000) {
        changed = false;
        iterations++;
        const conflicting = new Set<string>();
        for (const s of slots) {
          if (s.machineId === machineId) continue;
          if (candidate < s.end && candidate + duration > s.start) conflicting.add(s.machineId);
        }
        if (conflicting.size >= maxInst) {
          let minEnd = Infinity;
          for (const s of slots) {
            if (s.machineId === machineId) continue;
            if (candidate < s.end && candidate + duration > s.start) {
              minEnd = Math.min(minEnd, s.end);
            }
          }
          candidate = minEnd;
          changed = true;
        }
      }
      return candidate + duration <= shiftEnd ? candidate : -1;
    },

    /**
     * Full check combining isAvailable + findNextAvailable with detailed info.
     *
     * @param toolId    - Tool code
     * @param start     - Desired start (absolute minutes)
     * @param duration  - Duration in minutes
     * @param shiftEnd  - Shift end (absolute minutes)
     * @param machineId - Machine requesting the tool
     * @param instances - Number of physical copies (default 1)
     * @returns ToolCheckResult with full availability details
     */
    check(
      toolId: string,
      start: number,
      duration: number,
      shiftEnd: number,
      machineId: string,
      instances?: number,
    ): ToolCheckResult {
      const slots = timelines[toolId];
      const maxInst = instances ?? 1;
      const conflictingMachines: string[] = [];

      if (slots) {
        const seen = new Set<string>();
        for (const s of slots) {
          if (s.machineId === machineId) continue;
          if (start < s.end && start + duration > s.start) {
            if (!seen.has(s.machineId)) {
              seen.add(s.machineId);
              conflictingMachines.push(s.machineId);
            }
          }
        }
      }

      const isAvail = conflictingMachines.length < maxInst;
      const availableAt = isAvail
        ? start
        : this.findNextAvailable(toolId, start, duration, shiftEnd, machineId, instances);

      return {
        isAvailable: isAvail,
        availableAt,
        conflictCount: conflictingMachines.length,
        conflictingMachines,
      };
    },

    /**
     * Book a tool usage on a machine.
     *
     * @param toolId    - Tool code
     * @param start     - Absolute start minute
     * @param end       - Absolute end minute
     * @param machineId - Machine using the tool
     */
    book(toolId: string, start: number, end: number, machineId: string): void {
      if (!timelines[toolId]) timelines[toolId] = [];
      timelines[toolId].push({ start, end, machineId });
    },

    /**
     * Get all bookings for a specific tool (read-only copy).
     *
     * @param toolId - Tool code
     * @returns Array of booked slots
     */
    getBookings(toolId: string): ReadonlyArray<Readonly<ToolSlot>> {
      return timelines[toolId] ? [...timelines[toolId]] : [];
    },

    /** Clear all timelines (for test or reset) */
    clear(): void {
      for (const key of Object.keys(timelines)) {
        delete timelines[key];
      }
    },
  };
}

/** The return type of createToolTimeline() */
export type ToolTimeline = ReturnType<typeof createToolTimeline>;

File: apps/backend/src/domain/scheduling/constraints/tool_timeline.py
"""ToolTimeline constraint — ferramenta em 1 máquina de cada vez.

Port of constraints/tool-timeline.ts.
Same-machine reuse is OK (no self-conflict).
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ToolBooking:
    start: int
    end: int
    machine_id: str


@dataclass
class ToolCheckResult:
    is_available: bool
    available_at: int
    conflict_count: int = 0
    conflicting_machines: list[str] | None = None


class ToolTimeline:
    """Physical tool constraint: 1 tool instance per machine at a time."""

    def __init__(self) -> None:
        self._bookings: dict[str, list[ToolBooking]] = {}

    def is_available(self, tool_id: str, start: int, end: int, machine_id: str) -> bool:
        """Check if tool is available at [start, end) on machine_id."""
        for bk in self._bookings.get(tool_id, []):
            if bk.machine_id == machine_id:
                continue  # same machine OK
            if start < bk.end and end > bk.start:
                return False
        return True

    def find_next_available(
        self,
        tool_id: str,
        earliest: int,
        duration: int,
        shift_end: int,
        machine_id: str,
    ) -> int:
        """Find earliest start >= earliest where tool is free for duration."""
        candidate = earliest
        changed = True
        while changed:
            changed = False
            for bk in self._bookings.get(tool_id, []):
                if bk.machine_id == machine_id:
                    continue
                if candidate < bk.end and candidate + duration > bk.start:
                    candidate = bk.end
                    changed = True
        if candidate + duration > shift_end:
            return -1
        return candidate

    def check(
        self,
        tool_id: str,
        start: int,
        end: int,
        machine_id: str,
    ) -> ToolCheckResult:
        """Full check with conflict details."""
        conflicts: list[str] = []
        latest_end = start
        for bk in self._bookings.get(tool_id, []):
            if bk.machine_id == machine_id:
                continue
            if start < bk.end and end > bk.start:
                conflicts.append(bk.machine_id)
                latest_end = max(latest_end, bk.end)
        if conflicts:
            return ToolCheckResult(
                is_available=False,
                available_at=latest_end,
                conflict_count=len(conflicts),
                conflicting_machines=conflicts,
            )
        return ToolCheckResult(is_available=True, available_at=start)

    def book(self, tool_id: str, start: int, end: int, machine_id: str) -> None:
        self._bookings.setdefault(tool_id, []).append(
            ToolBooking(start=start, end=end, machine_id=machine_id)
        )

    def get_bookings(self, tool_id: str) -> list[ToolBooking]:
        return list(self._bookings.get(tool_id, []))

    def clear(self) -> None:
        self._bookings.clear()
